osutil/sync_dir.go

```diff
@@ -0,0 +1,118 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// FileState describes the expected content and meta data of a single file.
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+}
+
+var errSameState = fmt.Errorf("file state has not changed")
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions are what is desired. Unexpected files are
+// removed. Missing files are created and differing files are corrected.  Files
+// not matching the pattern are ignored.
+//
+// Note, that EnsureDirState() only checks for permissions and content. Other
+// security mechanisms, including file ownership and extended attributes are
+// *not* supported.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
+// returned to the caller for any extra processing that might be required (e.g.
+// to run some helper program).
+func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, removed []string, err error) {
+	if _, err := filepath.Match(glob, "foo"); err != nil {
+		panic(fmt.Sprintf("EnsureDirState got invalid pattern %q: %s", glob, err))
+	}
```

Comment:
This check is non-obvious perhaps since it seems Go isn't providing a way to compile a pattern. It would perhaps be more clear if you checked for ErrBadPattern which according to https://golang.org/pkg/path/filepath/#Match. I would have expected the bad pattern check to happen below with something like:

    if ok, err := filepath.Match(glob, baseName); !ok {
        ...
    }
    if err == ErrBadPattern {
        panic()
    }

If you don't like this, can you at least add a comment to make it clear what you are trying to achieve and why you are doing it where you are?